Add optional padding to the image

Damon3000s opened the pull request:
- Padding will NOT result in the output image being bigger. Instead the image will be inset by the amount of padding added.
- Added validation to ensure we don't try to add so much padding that the image is not visible and we do potentially invalid image operations

Damon3000s committed:
Add optional padding to the image

- Padding will NOT result in the output image being bigger. Instead the image will be inset by the amount of padding added.
- Added validation to ensure we don't try to add so much padding that the image is not visible and we do potentially invalid image operations

matt-edmondson commented on IconHelper/IconHelper.cs:
```diff
@@ -107,7 +113,13 @@ private static void Run(Arguments args)
 				if (newSize > args.Size)
 				{
-					image.Mutate(x => x.Resize(args.Size, args.Size));
+					image.Mutate(x => x.Resize(args.Size - (args.Padding * 2), args.Size - (args.Padding * 2))
```
take a local ie: `var paddedSize = args.Size - (args.Padding * 2);`

and use it for the duplicated params.

matt-edmondson commented on IconHelper/IconHelper.cs:
```diff
@@ -107,7 +113,13 @@ private static void Run(Arguments args)
+				}
+				else
+				{
+					image.Mutate(x => x.Resize(newSize - (args.Padding * 2), newSize - (args.Padding * 2))
```
same here

matt-edmondson replied:
Also try to decide the size and then perform the mutation, rather than having the mutation called twice with different sizes.

Damon3000s replied:
I don't think I can remove the else case. 

Due to the first case where we want to resize the image (eg. 512x512 -> 128x128) we don't want to add padding first because that will add the padding at 512x512 before it is resized, resulting in the padding not being the number of pixels we want (eg. 1px of padding at 512x512 when scaled down to 128x128 will in theory be 25% the size, and if another image is scaled from 256x256 to 128x128 it will be 50% the size and inconsistent to the other image).

It is very unfortunate and annoying that both cases now have 2 mutate steps, but it is the only way I could see to make the padding consistent regardless of the input image size and output image size (at least, not without significantly complicating the logic)

matt-edmondson replied:
Something like this:
```
int paddedSize = Math.Min(newSize, args.Size);
int contentSize = paddedSize - (args.Padding * 2);

image.Mutate(x => x.Resize(contentSize, contentSize)
	.Pad(paddedSize, paddedSize, Rgba32.ParseHex("00000000")));
```

Damon3000s replied:
Feedback discussed and understood.

And actioned.

Damon3000s committed:
Review feedback

Pull the padded size into a local variable

Damon3000s committed:
Delete accidentally added file

Damon3000s committed:
Review feedback

Collapse the if/else and do the additional resize-pad within the single mutate command

matt-edmondson commented on IconHelper/IconHelper.cs:
```diff
@@ -101,6 +101,12 @@ private static void Run(Arguments args)
 
+				// We intentionally only shrink the image and not grow it
+				bool shrinkImage = newSize > args.Size;
+				int finalSize = (newSize > args.Size) ? args.Size : newSize;
```
Math.Min

matt-edmondson commented on IconHelper/IconHelper.cs:
```diff
@@ -101,6 +101,12 @@ private static void Run(Arguments args)
+				// We intentionally only shrink the image and not grow it
+				bool shrinkImage = newSize > args.Size;
+				int finalSize = (newSize > args.Size) ? args.Size : newSize;
+				int sizeWithPadding = finalSize - (args.Padding * 2);
```
int contentSize

Damon3000s committed:
Review feedback

Rename variables for clarity

Damon3000s committed:
Review feedback

Remove unused variable
